### PR TITLE
[Dy2Stat]Support LoDTensorArray for slice op

### PR DESCRIPTION
**Pull request by liym27**
1. Support the input of slice is LoDTensorArray;
2. Support read elements of list in dygraph_to_static. For example:
```Python
arr = fluid.layers.array_write(x, i)
res1 = arr[i]
res2 = arr[0:i]
```

Note: https://github.com/PaddlePaddle/Paddle/pull/24250  (Support list pop in dy2stat) depends on this PR.

### CODE REVIEW

**Review by zhhsplendid — COMMENTED** [review]
LGTM within my knowledge

File: paddle/fluid/operators/slice_op.cc [hunk 1]
```diff
@@ -33,7 +33,13 @@ class SliceOp : public framework::OperatorWithKernel {
 
     PADDLE_ENFORCE_EQ(ctx->HasOutput("Out"), true,
                       "Output (Out) of slice op should not be null.");
-
+    auto x_var_type = ctx->GetInputsVarType("Input")[0];
+    if (x_var_type == framework::proto::VarType::LOD_TENSOR_ARRAY &&
+        ctx->IsRuntime()) {
```
Comment (zhiqiu) [inline paddle/fluid/operators/slice_op.cc]: A little concern, what about compile-time infershape?

File: paddle/fluid/operators/slice_op.cc [hunk 2]
```diff
@@ -34,18 +34,31 @@ class SliceOp : public framework::OperatorWithKernel {
...
+                            "The size of axes must be 1 when the Input of "
+                            "SliceOp is LoDTensorArray, "
+                            "but received %d.",
+                            axes.size()));
+      if (ctx->IsRuntime()) {
+        // If the var type of input is LOD_TENSOR_ARRAY,
+        // the output shape is determined by SliceKernel:Compute in runtime.
+        return;
+      } else {
+        // NOTE: A better way is needed to get accurate dims of tensor array.
```
Comment (zhiqiu) [inline paddle/fluid/operators/slice_op.cc:50]: Not important. Better add your name in NOTE, so other RDs can discuss with you if they are confusing about it.

**Review by zhiqiu — COMMENTED** [review]
LGTM

File: paddle/fluid/operators/slice_op.cc (hunk 1 above)
Comment (liym27) [reply paddle/fluid/operators/slice_op.cc]: > A little concern, what about compile-time infershape?

Done

File: paddle/fluid/operators/slice_op.cc (hunk 2 above)
Comment (liym27) [reply paddle/fluid/operators/slice_op.cc:50]: > Not important. Better add your name in NOTE, so other RDs can discuss with you if they are confusing about it.

Thanks. I will add my name in #24250

**Comment by Thunderbrook** [comment]
LGTM

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]

**Pull request merged**